src/csharp/Grpc.Core/Interceptors/Interceptor.cs

```diff
@@ -0,0 +1,406 @@
+#region Copyright notice and license
+
+// Copyright 2018 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Reflection;
+using System.Threading.Tasks;
+using Grpc.Core.Internal;
+
+namespace Grpc.Core.Interceptors
+{
+    /// <summary>
+    /// Serves as the base class for gRPC interceptors.
+    /// This is an EXPERIMENTAL API.
+    /// </summary>
+    public abstract class Interceptor
```

Comment:
Because it is an abstract class rather than interface, I deliberately decided to merge the two interfaces into one. The overhead is minimal and it will allow one class to implement both sides, which is nice, if you are a library author for example, and provide a tracing or logging interceptor, you expose one class to the user that they register on both sides. If you want, you can implement only one side, but if we went with two base classes, there was no way to consolidate them into one implementation.